### PR TITLE
Initial version of the vert.x 3.x instrumentation

### PR DESCRIPTION
**Pull request by manuel-alvarez-alvarez**
# What Does This Do
Includes the required instrumentation points to add support for IAST in vert.x 3.x

# Motivation

# Additional Notes

### CODE REVIEW

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/propagation/PropagationModuleImpl.java [hunk 1]
```diff
@@ -36,46 +39,61 @@ public void taintIfInputIsTainted(@Nonnull Object toTaint, @Nullable String inpu
...
+    if (ranges != null && ranges.length != 0) {
+      taintedObjects.taint(toTaint, Ranges.forObject(ranges[0].getSource()));
+    }
+  }
+
+  private static Range[] getRanges(final TaintedObjects taintedObjects, final Object input) {
+    if (input instanceof Taintable) {
+      final Taintable taintable = (Taintable) input;
+      final byte source = taintable.$$DD$getTaintedSource();
+      return source == SourceTypes.NONE ? null : Ranges.forObject(new Source(source, null, null));
```
[the remark this answers is not in the document]
Comment (manuel-alvarez-alvarez) [reply dd-java-agent/agent-iast/src/main/java/com/datadog/iast/propagation/PropagationModuleImpl.java:93]: Yep, you are right, at some point I event prepared an implementation with the whole `Range[]` I'll change it to propagate just the `Source`

File: dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/callsite/Servlet3RequestCallSite.java [hunk 2]
```diff
@@ -22,12 +22,7 @@ public static java.util.Map<java.lang.String, java.lang.String[]> afterGetParame
     final WebModule module = InstrumentationBridge.WEB;
     if (module != null) {
       try {
-        for (Map.Entry<String, String[]> entry : map.entrySet()) {
-          module.onParameterName(entry.getKey());
-          for (String value : entry.getValue()) {
-            module.onParameterValue(entry.getKey(), value);
-          }
-        }
+        module.onParameterValues(map);
```
Comment (smola) [inline dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/callsite/Servlet3RequestCallSite.java:25]: Is this related to Vert.x in some way?

File: dd-java-agent/instrumentation/vertx-web-3.5/src/main/java/datadog/trace/instrumentation/vertx_3_5/core/VertxHttpHeadersInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+public class VertxHttpHeadersInstrumentation extends Instrumenter.Iast
+    implements Instrumenter.ForSingleType {
+
+  public static final Reference VERTX_HTTP_HEADERS =
+      new Reference.Builder("io.vertx.core.http.impl.headers.VertxHttpHeaders").build();
+
+  private final String className = VertxHttpHeadersInstrumentation.class.getName();
+
+  public VertxHttpHeadersInstrumentation() {
+    super("vertx", "vertx-3.5");
```
Comment (smola) [inline dd-java-agent/instrumentation/vertx-web-3.5/src/main/java/datadog/trace/instrumentation/vertx_3_5/core/VertxHttpHeadersInstrumentation.java:32]: Maybe we should also add `vertx-3` or `vertx-3.x` as second parameter? (I'm not sure if it's used in other modules?)

File: dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/callsite/Servlet3RequestCallSite.java (hunk 2 above)
Comment (manuel-alvarez-alvarez) [reply dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/callsite/Servlet3RequestCallSite.java]: Nope, I removed all the unrelated changes

File: dd-java-agent/instrumentation/vertx-web-3.5/src/main/java/datadog/trace/instrumentation/vertx_3_5/core/VertxHttpHeadersInstrumentation.java (hunk 3 above)
Comment (manuel-alvarez-alvarez) [reply dd-java-agent/instrumentation/vertx-web-3.5/src/main/java/datadog/trace/instrumentation/vertx_3_5/core/VertxHttpHeadersInstrumentation.java:32]: Other modules are defined as:

```java
  public RoutingContextImplInstrumentation() {
    super("vertx", "vertx-3.4");
  }
```

so I'm not sure if we should be more specific in here.

**Review by smola — APPROVED** [review]

**Pull request merged**